File: apps/sqs/java/src/main/java/example/Producer.java
package example;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.lambda.runtime.events.APIGatewayV2HTTPEvent;
import com.amazonaws.services.lambda.runtime.events.APIGatewayV2HTTPResponse;
import com.amazonaws.services.sqs.AmazonSQS;
import com.amazonaws.services.sqs.AmazonSQSClientBuilder;
import com.amazonaws.services.sqs.model.SendMessageRequest;
import org.json.JSONObject;

import static example.Context.currentTraceId;
import static example.Context.runtime;

public class Producer implements RequestHandler<APIGatewayV2HTTPEvent, APIGatewayV2HTTPResponse>{

  private static final AmazonSQS client = AmazonSQSClientBuilder.defaultClient();
  private static final String[] queueUrls = System.getenv("SQS_QUEUE_URLS").split(",");

  @Override
  public APIGatewayV2HTTPResponse handleRequest(APIGatewayV2HTTPEvent event, Context context)
  {
    final String msg = new JSONObject()
      .put("runtime", runtime)
      .put("trace_id", currentTraceId())
      .toString();

    for (String url : queueUrls) {
        System.out.println("sending sqs message " + msg + " to " + url);
        client.sendMessage(new SendMessageRequest()
          .withQueueUrl(url)
          .withMessageBody(msg)
        );
    }

    APIGatewayV2HTTPResponse response = new APIGatewayV2HTTPResponse();
    response.setStatusCode(200);
    response.setBody("ok");
    return response;
  }
}

File: apps/sqs/golang/producer/producer.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strings"

	ddlambda "github.com/DataDog/datadog-lambda-go"
	"github.com/DataDog/sample-apps/apps/sqs/golang/internal"
	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/sqs"
	"github.com/aws/aws-sdk-go/aws"
	awstrace "gopkg.in/DataDog/dd-trace-go.v1/contrib/aws/aws-sdk-go-v2/aws"
)

var (
	client = func() *sqs.Client {
		cfg, _ := config.LoadDefaultConfig(context.Background())
		awstrace.AppendMiddleware(&cfg)
		return sqs.NewFromConfig(cfg)
	}()

	queueUrls = strings.Split(os.Getenv("SQS_QUEUE_URLS"), ",")
)

func producer(ctx context.Context, event events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
	msg, _ := json.Marshal(internal.Message{
		Runtime: internal.Runtime,
		TraceID: internal.TraceID(ctx),
	})

	for _, url := range queueUrls {
		fmt.Printf("sending sqs message %s to %s\n", string(msg), url)
		client.SendMessage(ctx, &sqs.SendMessageInput{
			MessageBody: aws.String(string(msg)),
			QueueUrl:    aws.String(url),
		})
	}

	return events.APIGatewayProxyResponse{
		Body:       "ok",
		StatusCode: 200,
	}, nil
}

func main() {
	lambda.Start(ddlambda.WrapFunction(producer, nil))
}
